### PR TITLE
Disable unnecessary default dependencies of Eigen

### PR DESCRIPTION
**Pull request by HadrienG2**
Considering how few mandatory dependencies Eigen has, it is a shame that Spack's Eigen package pulls in so many barely relevant dependencies by default, and requires users who care about their Docker image sizes and package build times to disable all of them one by one.

I would therefore propose to change the default to a more bare-bones Eigen install, and let those who are interested in Eigen's integration with various optional dependencies ask for them using the relevant variants.

### CODE REVIEW

**Comment by citibeth** [comment]
Can you summarize what these options do?  I'm not sure they're all about integration.  For example, what is the consequence of disabling FFTW?

**Comment by HadrienG2** [comment]
Here is what I found by grepping through the Eigen codebases for these dependencies:

# Metis

[Eigen can solve sparse linear systems in various ways](http://eigen.tuxfamily.org/dox/group__TopicSparseSystems.html). Some direct solvers work by factoring the sparse matrix. Before doing so, nonzero matrix elements may be reordered as a preprocessing step so that factoring the matrix creates less fill-in. Several reordering algorithms are supported, some of which are built-in. The METIS algorithm, however, is implemented externally by the eponymous library.

If this dependency is not provided, then Eigen will not be able to reorder the matrix in this specific way during resolution of sparse linear systems. Other reordering methods remain available as an alternative.

# Pastix

Like metis, this dependency is used exclusively for the purpose of sparse linear system resolution. It works at a higher level by providing a set of sparse matrix factorization algorithms (LLt, LDLt, LU). All of these factorization algorithms also have a built-in implementation in Eigen, but the pastix implementations are multi-threaded, and said to be "optimized for tough problems and symmetric patterns".

If this dependency is not available, these implementations of sparse matrix factorization will not be available. The built-in Eigen implementations of these algorithms remain available, although they may be less efficient, so there is no major loss of functionality.

# Scotch

This dependency is never user directly by Eigen, but only requested as a plug-in to pastix during CMake configuration of pastix-related Eigen tests and benchmarks. The pastix documentation suggests that it plays a role similar to that of metis, implementing a kind of sparse matrix reordering as a preprocessing step to sparse matrix factorization that may improve memory usage or performance.

# FFTW

This dependency is not used by the official Eigen library, but only by a plug-in that is included in the Eigen source package but [comes with no official support guarantee from the Eigen team](https://github.com/eigenteam/eigen-git-mirror/blob/master/unsupported/README.txt). The relevant plug-in implements fast Fourier transforms using several different backends. Use of the fftw backend requires the fftw library to be available.

If fftw is not available, then this back-end cannot be used with that plug-in. A lighter-weight back-end called "kissfft" is also provided in the unsupported section of the Eigen source tree, so no net functionality is lost in this case, only some room for performance optimization.

# SuiteSparse

This dependency is used for sparse system resolution. It provides extra factorization algorithms which are not provided by the base Eigen package. These extra algorithms do not enable the resolution of new kinds of sparse systems, but presumably, like those of pastix, they solve the same systems more efficiently.

# MPFR

This dependency is used by an unsupported Eigen plugin to implement support for bignum arithmetic as an alternative to IEEE arithmetic. If MPFR is unavailable, then this plugin cannot be used.

----

As you can see, none of these dependencies are needed when Eigen is used, as happens most frequently, for the purpose of computing dense IEEE algebra. Most of them enable different (presumably more efficient) ways to resolve sparse linear systems in addition to those built into Eigen, while others are used in unsupported plugins for the purpose of adding new FFT and bignum capabilities on top of the base Eigen library.

**Comment by citibeth** [comment]
Thank you for the research.  Except for MPFR, it looks like these are all related to sparse matrices; which are of course only used by a subset of Eigen users.  Similarly, MPFR is going to be used by cryptographers and such, but not by your typical Eigen user solving real-valued matrices.  In addition, it looks like Eigen hasn't really settled on which set of back-ends they want to use or think is best, so they're "supporting" all of them in experimental mode.

I agree with turning these all off by default.  However, I would like to see some summary of what we discovered here put into the description strings for the variants, in the `package.py`, or both.  Ideally, users should be able to make a good decision on these dependencies without having to dig past `package.py`.

**Comment by HadrienG2** [comment]
Here is an attempt at clarifying things a bit while keeping the description short. Hopefully the CI will not complain about that 81 char line, if it does I'll find a way to save one character for punch card compatibility... :wink:

**Comment by HadrienG2** [comment]
...done. Punch card compatibility checker should be happy now.
